// carbon-addons-ics/es/components/Notification/Notification.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlineNotification = exports.ToastNotification = exports.NotificationTextDetails = exports.NotificationButton = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icon = _interopRequireDefault(require("../Icon"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var NotificationButton =
/*#__PURE__*/
function (_Component) {
  _inherits(NotificationButton, _Component);

  function NotificationButton() {
    _classCallCheck(this, NotificationButton);

    return _possibleConstructorReturn(this, _getPrototypeOf(NotificationButton).apply(this, arguments));
  }

  _createClass(NotificationButton, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          ariaLabel = _this$props.ariaLabel,
          className = _this$props.className,
          iconDescription = _this$props.iconDescription,
          name = _this$props.name,
          rest = _objectWithoutProperties(_this$props, ["ariaLabel", "className", "iconDescription", "name"]);

      var classes = (0, _classnames.default)('bx--toast-notification__close-button', className);
      return _react.default.createElement("button", _extends({}, rest, {
        type: "button",
        className: classes
      }), _react.default.createElement(_Icon.default, {
        description: iconDescription,
        className: "bx--toast-notification__close-icon",
        "aria-label": ariaLabel,
        name: name
      }));
    }
  }]);

  return NotificationButton;
}(_react.Component);

exports.NotificationButton = NotificationButton;

_defineProperty(NotificationButton, "propTypes", {
  className: _propTypes.default.string,
  ariaLabel: _propTypes.default.string,
  type: _propTypes.default.string,
  iconDescription: _propTypes.default.string,
  name: _propTypes.default.string
});

_defineProperty(NotificationButton, "defaultProps", {
  ariaLabel: 'close notification',
  iconDescription: 'close icon',
  name: 'close'
});

var NotificationTextDetails =
/*#__PURE__*/
function (_Component2) {
  _inherits(NotificationTextDetails, _Component2);

  function NotificationTextDetails() {
    _classCallCheck(this, NotificationTextDetails);

    return _possibleConstructorReturn(this, _getPrototypeOf(NotificationTextDetails).apply(this, arguments));
  }

  _createClass(NotificationTextDetails, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          title = _this$props2.title,
          subtitle = _this$props2.subtitle,
          link = _this$props2.link,
          linkTitle = _this$props2.linkTitle,
          className = _this$props2.className,
          notificationType = _this$props2.notificationType,
          kind = _this$props2.kind,
          rest = _objectWithoutProperties(_this$props2, ["title", "subtitle", "link", "linkTitle", "className", "notificationType", "kind"]);

      if (notificationType === 'toast') {
        return _react.default.createElement("div", _extends({}, rest, {
          className: "bx--toast-notification__details ".concat(className)
        }), _react.default.createElement(_Icon.default, {
          description: title,
          className: "bx--toast-notification__icon",
          "aria-label": title,
          name: "".concat(kind, "--glyph")
        }), _react.default.createElement("div", {
          className: "bx--toast-notification__text-wrapper"
        }, _react.default.createElement("div", {
          className: "bx--toast-notification__title"
        }, title), _react.default.createElement("div", {
          className: "bx--toast-notification__subtitle"
        }, subtitle), _react.default.createElement("a", {
          className: "bx--toast-notification__link",
          href: link
        }, linkTitle)));
      }

      return _react.default.createElement("div", _extends({}, rest, {
        className: "bx--inline-notification__text-wrapper"
      }), _react.default.createElement("div", {
        className: "bx--inline-notification__title"
      }, title), _react.default.createElement("div", {
        className: "bx--inline-notification__subtitle"
      }, subtitle));
    }
  }]);

  return NotificationTextDetails;
}(_react.Component);

exports.NotificationTextDetails = NotificationTextDetails;

_defineProperty(NotificationTextDetails, "propTypes", {
  className: _propTypes.default.string,
  title: _propTypes.default.string,
  subtitle: _propTypes.default.node,
  link: _propTypes.default.string,
  linkTitle: _propTypes.default.string,
  notificationType: _propTypes.default.oneOf(['toast', 'inline']),
  kind: _propTypes.default.oneOf(['error', 'info', 'success', 'warning']).isRequired
});

_defineProperty(NotificationTextDetails, "defaultProps", {
  title: 'title',
  subtitle: 'subtitle',
  link: '#',
  linkTitle: 'Undo?',
  notificationType: 'toast',
  kind: 'success'
});

var ToastNotification =
/*#__PURE__*/
function (_Component3) {
  _inherits(ToastNotification, _Component3);

  function ToastNotification(props) {
    var _this;

    _classCallCheck(this, ToastNotification);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ToastNotification).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleCloseButtonClick", function (evt) {
      _this.setState({
        open: false
      });

      _this.props.onCloseButtonClick(evt);
    });

    _this.state = {
      open: true
    };
    return _this;
  }

  _createClass(ToastNotification, [{
    key: "render",
    value: function render() {
      if (!this.state.open) {
        return null;
      }

      var _this$props3 = this.props,
          role = _this$props3.role,
          notificationType = _this$props3.notificationType,
          onCloseButtonClick = _this$props3.onCloseButtonClick,
          iconDescription = _this$props3.iconDescription,
          className = _this$props3.className,
          link = _this$props3.link,
          linkTitle = _this$props3.linkTitle,
          subtitle = _this$props3.subtitle,
          title = _this$props3.title,
          kind = _this$props3.kind,
          rest = _objectWithoutProperties(_this$props3, ["role", "notificationType", "onCloseButtonClick", "iconDescription", "className", "link", "linkTitle", "subtitle", "title", "kind"]);

      var classes = (0, _classnames.default)('bx--toast-notification', _defineProperty({}, "bx--toast-notification--".concat(this.props.kind), this.props.kind), className);
      return _react.default.createElement("div", _extends({}, rest, {
        role: role,
        kind: kind,
        className: classes
      }), _react.default.createElement(NotificationTextDetails, {
        title: title,
        kind: kind,
        subtitle: subtitle,
        link: link,
        linkTitle: linkTitle,
        notificationType: notificationType
      }), _react.default.createElement(NotificationButton, {
        iconDescription: iconDescription,
        onClick: this.handleCloseButtonClick
      }));
    }
  }]);

  return ToastNotification;
}(_react.Component);

exports.ToastNotification = ToastNotification;

_defineProperty(ToastNotification, "propTypes", {
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  kind: _propTypes.default.oneOf(['error', 'info', 'success', 'warning']).isRequired,
  title: _propTypes.default.string.isRequired,
  subtitle: _propTypes.default.node.isRequired,
  link: _propTypes.default.string,
  linkTitle: _propTypes.default.string,
  role: _propTypes.default.string.isRequired,
  onCloseButtonClick: _propTypes.default.func,
  iconDescription: _propTypes.default.string.isRequired,
  notificationType: _propTypes.default.string
});

_defineProperty(ToastNotification, "defaultProps", {
  kind: 'error',
  title: 'provide a title',
  subtitle: 'provide a subtitle',
  link: '#',
  linkTitle: 'Undo?',
  role: 'alert',
  notificationType: 'toast',
  iconDescription: 'closes notification',
  onCloseButtonClick: function onCloseButtonClick() {}
});

var InlineNotification =
/*#__PURE__*/
function (_Component4) {
  _inherits(InlineNotification, _Component4);

  function InlineNotification() {
    _classCallCheck(this, InlineNotification);

    return _possibleConstructorReturn(this, _getPrototypeOf(InlineNotification).apply(this, arguments));
  }

  _createClass(InlineNotification, [{
    key: "render",
    value: function render() {
      var _this$props4 = this.props,
          role = _this$props4.role,
          notificationType = _this$props4.notificationType,
          iconDescription = _this$props4.iconDescription,
          className = _this$props4.className,
          subtitle = _this$props4.subtitle,
          title = _this$props4.title,
          kind = _this$props4.kind,
          rest = _objectWithoutProperties(_this$props4, ["role", "notificationType", "iconDescription", "className", "subtitle", "title", "kind"]);

      var classes = (0, _classnames.default)('bx--inline-notification', _defineProperty({}, "bx--inline-notification--".concat(this.props.kind), this.props.kind), className);
      return _react.default.createElement("div", _extends({}, rest, {
        role: role,
        kind: kind,
        className: classes
      }), _react.default.createElement("div", {
        className: "bx--inline-notification__details"
      }, _react.default.createElement(_Icon.default, {
        description: iconDescription,
        className: "bx--inline-notification__icon",
        "aria-label": "close",
        name: "".concat(kind, "--glyph")
      }), _react.default.createElement(NotificationTextDetails, {
        title: title,
        subtitle: subtitle,
        notificationType: notificationType
      })));
    }
  }]);

  return InlineNotification;
}(_react.Component);

exports.InlineNotification = InlineNotification;

_defineProperty(InlineNotification, "propTypes", {
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  kind: _propTypes.default.oneOf(['error', 'info', 'success', 'warning']).isRequired,
  title: _propTypes.default.string.isRequired,
  subtitle: _propTypes.default.node.isRequired,
  role: _propTypes.default.string.isRequired,
  iconDescription: _propTypes.default.string.isRequired,
  notificationType: _propTypes.default.string
});

_defineProperty(InlineNotification, "defaultProps", {
  role: 'alert',
  notificationType: 'inline',
  iconDescription: 'closes notification'
});
