// carbon-addons-ics/lib/components/ModalWrapper/ModalWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Modal = _interopRequireDefault(require("../Modal"));

var _carbonComponentsReact = require("carbon-components-react");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ModalWrapper =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ModalWrapper, _React$Component);

  function ModalWrapper(props) {
    var _this;

    _classCallCheck(this, ModalWrapper);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ModalWrapper).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleOpen", function () {
      _this.setState({
        open: true
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleClose", function () {
      _this.props.onRequestClose();

      _this.setState({
        open: false
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleOnRequestSubmit", function () {
      var _this$props = _this.props,
          handleSubmit = _this$props.handleSubmit,
          shouldCloseAfterSubmit = _this$props.shouldCloseAfterSubmit;

      if (handleSubmit()) {
        if (shouldCloseAfterSubmit) {
          _this.handleClose();
        }
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleKeyDown", function (evt) {
      if (evt.which === 27) {
        _this.handleClose();

        _this.props.onKeyDown(evt);
      }
    });

    _this.state = {
      open: false
    };
    return _this;
  }

  _createClass(ModalWrapper, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          id = _this$props2.id,
          buttonTriggerText = _this$props2.buttonTriggerText,
          triggerButtonkind = _this$props2.triggerButtonkind,
          modalHeading = _this$props2.modalHeading,
          passiveModal = _this$props2.passiveModal,
          primaryButtonText = _this$props2.primaryButtonText,
          secondaryButtonText = _this$props2.secondaryButtonText,
          handleSubmit = _this$props2.handleSubmit,
          disabled = _this$props2.disabled,
          onKeyDown = _this$props2.onKeyDown,
          iconDescription = _this$props2.iconDescription,
          primaryButtonDisabled = _this$props2.primaryButtonDisabled,
          onSecondarySubmit = _this$props2.onSecondarySubmit;
      var props = {
        id: id,
        modalHeading: modalHeading,
        passiveModal: passiveModal,
        primaryButtonText: primaryButtonText,
        secondaryButtonText: secondaryButtonText,
        open: this.state.open,
        onRequestClose: this.handleClose,
        onRequestSubmit: this.handleOnRequestSubmit,
        onKeyDown: onKeyDown,
        iconDescription: iconDescription,
        primaryButtonDisabled: primaryButtonDisabled,
        onSecondarySubmit: onSecondarySubmit
      };
      return _react.default.createElement("div", {
        role: "presentation",
        onKeyDown: this.handleKeyDown.bind(this)
      }, _react.default.createElement(_carbonComponentsReact.Button, {
        disabled: disabled,
        kind: triggerButtonkind,
        onClick: this.handleOpen
      }, buttonTriggerText), _react.default.createElement(_Modal.default, props, this.props.children));
    }
  }]);

  return ModalWrapper;
}(_react.default.Component);

exports.default = ModalWrapper;

_defineProperty(ModalWrapper, "propTypes", {
  status: _propTypes.default.string,
  handleOpen: _propTypes.default.func,
  children: _propTypes.default.node,
  buttonTriggerText: _propTypes.default.string,
  withHeader: _propTypes.default.bool,
  modalBeforeContent: _propTypes.default.bool,
  handleSubmit: _propTypes.default.func,
  disabled: _propTypes.default.bool,
  triggerButtonkind: _propTypes.default.oneOf(['primary', 'secondary', 'danger', 'ghost']),
  shouldCloseAfterSubmit: _propTypes.default.bool,
  // Modal Props
  id: _propTypes.default.string,
  modalHeading: _propTypes.default.string,
  passiveModal: _propTypes.default.bool,
  primaryButtonText: _propTypes.default.string,
  secondaryButtonText: _propTypes.default.string,
  onRequestClose: _propTypes.default.func,
  onRequestSubmit: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  iconDescription: _propTypes.default.string,
  primaryButtonDisabled: _propTypes.default.bool,
  onSecondarySubmit: _propTypes.default.func
});

_defineProperty(ModalWrapper, "defaultProps", {
  primaryButtonText: 'Save',
  secondaryButtonText: 'Cancel',
  triggerButtonkind: 'primary',
  disabled: false,
  onRequestClose: function onRequestClose() {},
  onKeyDown: function onKeyDown() {}
});
