// carbon-addons-ics/lib/components/CardContent/CardContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var CardContent = function CardContent(_ref) {
  var className = _ref.className,
      children = _ref.children,
      cardTitle = _ref.cardTitle,
      cardLink = _ref.cardLink,
      cardInfo = _ref.cardInfo,
      cardAvatar = _ref.cardAvatar,
      other = _objectWithoutProperties(_ref, ["className", "children", "cardTitle", "cardLink", "cardInfo", "cardAvatar"]);

  var cardContentClasses = (0, _classnames.default)(_defineProperty({
    'bx--card__card-overview': true,
    'bx--card__business-card': true
  }, className, className));
  var cardLinkContent = cardLink ? cardLink.map(function (link, key) {
    return _react.default.createElement("a", {
      key: key,
      href: link,
      className: "bx--about__title--link"
    }, link);
  }) : '';
  var cardInfoContent = cardInfo ? cardInfo.map(function (info, key) {
    return _react.default.createElement("p", {
      key: key,
      className: "bx--card__info"
    }, info);
  }) : '';
  var cardLinkContentArray = Object.keys(cardLinkContent);
  var cardInfoContentArray = Object.keys(cardInfoContent);
  return _react.default.createElement("div", _extends({}, other, {
    className: cardContentClasses
  }), children, _react.default.createElement("div", {
    className: "bx--card__business-card"
  }, cardAvatar && _react.default.createElement("div", {
    className: "bx--about__icon"
  }, _react.default.createElement("img", {
    src: cardAvatar,
    alt: "Avatar",
    className: "bx--about__icon--img"
  })), _react.default.createElement("div", {
    className: "bx--card__business-card__title"
  }, _react.default.createElement("p", {
    id: "card-app-title",
    className: "bx--about__title--name"
  }, cardTitle), cardLinkContentArray.map(function (info, key) {
    return cardLinkContent[key];
  }), cardInfoContentArray.map(function (info, key) {
    return cardInfoContent[key];
  }))));
};

CardContent.propTypes = {
  children: _propTypes.default.node,
  cardTitle: _propTypes.default.string,
  cardLink: _propTypes.default.node,
  cardInfo: _propTypes.default.array,
  cardAvatar: _propTypes.default.string,
  className: _propTypes.default.string
};
CardContent.defaultProps = {
  cardTitle: 'card title'
};
var _default = CardContent;
exports.default = _default;
