// carbon-addons-ics/lib/components/Tag/Tag.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _ActionIcon = _interopRequireDefault(require("../ActionIcon"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Tag =
/*#__PURE__*/
function (_Component) {
  _inherits(Tag, _Component);

  function Tag(props) {
    var _this;

    _classCallCheck(this, Tag);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Tag).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "buildAction", function () {
      var _this$props = _this.props,
          action = _this$props.action,
          actionDescription = _this$props.actionDescription,
          actionFunction = _this$props.actionFunction,
          onClick = _this$props.onClick;
      var iconFunction = actionFunction ? actionFunction : onClick;

      if (action) {
        var actionClasses = (0, _classnames.default)(_defineProperty({
          'bx--tag__left': true
        }, "bx--tag__left--".concat(action), true));
        return _react.default.createElement("div", {
          className: actionClasses
        }, _react.default.createElement(_ActionIcon.default, {
          rounded: true,
          icon: action === 'remove' ? 'error' : action,
          iconDescription: actionDescription,
          tabIndex: 0,
          onClick: action === 'remove' ? _this.removeTag : iconFunction
        }));
      }

      return;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "removeTag", function () {
      _this.setState({
        showTag: false
      });
    });

    _this.state = {
      showTag: true
    };
    return _this;
  }

  _createClass(Tag, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          actionFunction = _this$props2.actionFunction,
          actionDescription = _this$props2.actionDescription,
          children = _this$props2.children,
          className = _this$props2.className,
          style = _this$props2.style,
          role = _this$props2.role,
          tabindex = _this$props2.tabindex,
          remove = _this$props2.remove,
          rest = _objectWithoutProperties(_this$props2, ["actionFunction", "actionDescription", "children", "className", "style", "role", "tabindex", "remove"]);

      var tagClasses = (0, _classnames.default)(_defineProperty({
        'bx--tag': true,
        'bx--tag--remove': remove,
        'bx--tag--dark': style === 'dark',
        'bx--tag--light': style === 'light'
      }, className, className));
      var actionIcon = this.buildAction();
      return this.state.showTag ? _react.default.createElement("li", {
        className: "bx--tag__item"
      }, actionIcon, _react.default.createElement("span", _extends({
        className: tagClasses,
        tabIndex: tabindex,
        role: role
      }, rest), children, remove && _react.default.createElement("div", {
        className: "bx--tag__remove-icon"
      }, _react.default.createElement(_ActionIcon.default, {
        rounded: true,
        icon: "close",
        iconDescription: "remove tag",
        onClick: this.props.onClick || this.removeTag
      })))) : '';
    }
  }]);

  return Tag;
}(_react.Component);

exports.default = Tag;

_defineProperty(Tag, "propTypes", {
  action: _propTypes.default.string,
  actionDescription: _propTypes.default.string,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  onBlur: _propTypes.default.func,
  onClick: _propTypes.default.func,
  onFocus: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  onKeyUp: _propTypes.default.func,
  onMouseDown: _propTypes.default.func,
  onMouseEnter: _propTypes.default.func,
  onMouseLeave: _propTypes.default.func,
  onMouseUp: _propTypes.default.func,
  remove: _propTypes.default.bool,
  role: _propTypes.default.string,
  style: _propTypes.default.oneOf(['dark', 'light']),
  tabindex: _propTypes.default.string
});

_defineProperty(Tag, "defaultProps", {
  action: null,
  remove: false,
  role: 'tag',
  style: 'dark',
  tabindex: '0'
});
