// carbon-addons-ics/es/components/TextArea/TextArea.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Textarea = function Textarea(_ref) {
  var className = _ref.className,
      id = _ref.id,
      labelText = _ref.labelText,
      _onChange = _ref.onChange,
      _onClick = _ref.onClick,
      invalid = _ref.invalid,
      invalidText = _ref.invalidText,
      rest = _objectWithoutProperties(_ref, ["className", "id", "labelText", "onChange", "onClick", "invalid", "invalidText"]);

  var textareaProps = {
    id: id,
    onChange: function onChange(evt) {
      if (!rest.disabled) {
        _onChange(evt);
      }
    },
    onClick: function onClick(evt) {
      if (!rest.disabled) {
        _onClick(evt);
      }
    }
  };
  var textareaClasses = (0, _classnames.default)('bx--text-area', className);
  var label = labelText ? _react.default.createElement("label", {
    htmlFor: id,
    className: "bx--label"
  }, labelText) : null;
  var error = invalid ? _react.default.createElement("div", {
    className: "bx--form-requirement"
  }, invalidText) : null;
  var input = invalid ? _react.default.createElement("textarea", _extends({}, rest, textareaProps, {
    className: textareaClasses,
    "data-invalid": true
  })) : _react.default.createElement("textarea", _extends({}, rest, textareaProps, {
    className: textareaClasses
  }));

  var span = _react.default.createElement("span", {
    className: "bx--mi__underline"
  });

  return _react.default.createElement("div", {
    className: "bx--form-item bx--text-area__container"
  }, label, input, span, error);
};

Textarea.propTypes = {
  className: _propTypes.default.string,
  cols: _propTypes.default.number,
  disabled: _propTypes.default.bool,
  id: _propTypes.default.string,
  labelText: _propTypes.default.string,
  onChange: _propTypes.default.func,
  onClick: _propTypes.default.func,
  placeholder: _propTypes.default.string,
  rows: _propTypes.default.number,
  invalid: _propTypes.default.bool,
  invalidText: _propTypes.default.string
};
Textarea.defaultProps = {
  disabled: false,
  onChange: function onChange() {},
  onClick: function onClick() {},
  placeholder: 'Hint text here',
  rows: 4,
  cols: 50,
  invalid: false,
  labelText: 'Provide labelText',
  invalidText: 'Provide invalidText'
};
var _default = Textarea;
exports.default = _default;
